Library/Homebrew/formula_info.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+# Formula info drawn from an external `brew info --json` call
+
+class FormulaInfoFromJson
+  # The whole info structure parsed from the JSON
+  attr_accessor :info
+
+  def initialize(info)
+    @info = info
+  end
+
+  # Looks up formula on disk and reads its info.
+  # Returns nil if formula is absent or if there was an error reading it.
+  def self.lookup(name)
+    json = Utils.popen_read(HOMEBREW_BREW_FILE, "info", "--json=v1", name)
+
+    return unless $CHILD_STATUS.success?
+
+    Homebrew.force_utf8!(json)
+    FormulaInfoFromJson.new(JSON.parse(json)[0])
+  end
+
+  def bottle_tags
+    return [] unless info["bottle"]["stable"]
+
+    info["bottle"]["stable"]["files"].keys
+  end
+
+  def bottle_info(my_bottle_tag = Utils::Bottles.tag)
+    tag_s = my_bottle_tag.to_s
+    return unless info["bottle"]["stable"]
+
+    btl_info = info["bottle"]["stable"]["files"][tag_s]
+    return unless btl_info
+
+    BottleInfo.new(btl_info["url"], btl_info["sha256"])
+  end
+
+  def bottle_info_any
+    bottle_info(any_bottle_tag)
+  end
+
+  def any_bottle_tag
+    tag = Utils::Bottles.tag.to_s
+    # Prefer native bottles as a convenience for download caching
+    bottle_tags.include?(tag) ? tag : bottle_tags.first
+  end
+
+  def version(spec_type)
+    version_str = info["versions"][spec_type.to_s]
+    version_str && Version.create(version_str)
+  end
+
+  def pkg_version(spec_type = :stable)
+    PkgVersion.new(version(spec_type), revision)
+  end
+
+  def revision
+    info["revision"]
+  end
+end
+
+# Bottle info as used internally by pull, with alternate platform support.
+class BottleInfo
```

Comment:
This could live in `brew pull` I think; there's not enough here to be useful standalone.

Reply:
The BottleInfo class is only called by the formula_info class once. The comment is misleading (and I am not really sure it is helpful):

> # Bottle info as used internally by pull, with alternate platform support.

I removed that comment